### DISCUSSION TITLE
How to avoid propagating Futures and load value in main()?

### DISCUSSION THREAD

**Discussion by eximius313** [rrousselGit/riverpod#2984] [Q&A] [answered]
Currently I have such a code:
```dart
void main() async {
  final config = await File("settings.json").readAsString();
    runApp(
    MultiProvider(
      providers: [
        Provider<SimpleRepository>.value(value: SimpleRepository(config)),
        Provider<OterRepository>.value(value: OterRepository(config)),
        Provider<FooRepository>.value(value: FooRepository(config)),
      ],
      child: MyApp(),
    )
  );
}
```

As you can see, there is one - rather quick - async operation, and then all of the repositories can be acessed synchronously.
Unfortunatelly, when I want to use Riverpod in order to provide theese repositories, I need to pass them the config.
And here is where the problems begin:
1. When I want to make "configProvider":
```dart
final configProvider = FutureProvider<String>((ref) async => await File("settings.json").readAsString());
```
then Riverpod forces me to use FutureProvider, and then this future is propagated all the way down, because:
```dart
final simpleRepoProvider = FutureProvider(
    (ref) async => SimpleRepo(await ref.watch(configProvider.future)));
```
and then I must use FutureBuilders everywhere (although currently they are not required)

2. on the other hand, if I do:
```dart
final simpleRepoProvider = Provider.family<SimpleRepo, String>(
    (ref, config) => SimpleRepo(config));
```
then I must load the config everywhere (because I must pass it everywhere I use provider).

Is there other way around this problem?

Kind regards,

**Comment by AhmedLSayed9** [comment, marked as answer]
```dart
final configAsyncProvider = FutureProvider<String>((ref) async => await File("settings.json").readAsString());
final configProvider = Provider<String>((ref) => ref.watch(configAsyncProvider).requireValue);
```

at main before runApp, warm-up `configAsyncProvider`:
```dart
final container = ProviderContainer();
await container.read(configAsyncProvider.future);
runApp(
  ProviderScope(
    parent: container,
    child: const MyApp(),
  ),
);
```

then later, you can use `configProvider` to access the config synchronously.

**Reply by eximius313** [reply]
Thanks @AhmedLSayed9! This does the job!

**Comment by rrousselGit** [comment]
The general recommendations are here https://riverpod.dev/docs/essentials/eager_initialization

It's suggested to not do things in the main and instead move it to a consumer like in the docs. Then to use requireValue

**Reply by eximius313** [reply]
@rrousselGit, I've read documentation regarding Eager initialization and I'm using it in my project, but I'm not sure how is it related to my problem.
I have few Repositories that rely on loading the Config and I use them in dozen of places. I don't want to use `requireValue` in this dozen of places and thanks to @AhmedLSayed9 - I can use it only once.
Does Eager initialization provide better solution that I can not spot yet?

**Reply by AhmedLSayed9** [reply]
You can just do the same by moving the warm-up to a consumer instead of the main.

The idea is that warming-up in the main will prevent the app from loading if the future throws, but you can do some try/catch. Also, you should define your observers before the warm-up, something like:
```dart
final container = ProviderContainer(observers: [ProviderLogger()]);
try{
  await container.read(configAsyncProvider.future);
}catch(_){}
runApp(
  ProviderScope(
    parent: container,
    child: const MyApp(),
  ),
);
```

If you've some custom splash screen, you better move the warm-up there instead of the main during the native splash.

**Reply by rrousselGit** [reply]
Using requireValue is the recommended way

You're free to use overrides. But the page I linked explicitly discourages it in favor of using requireValue.

Trying to remove the requireValue is usually over engineering it.
